### ISSUE TITLE
After two objects collide, IDs are exchanged.

### ISSUE THREAD

**Issue by armine105** [tryolabs/norfair#288]
Hello
How are you?
I am using your method to do tracking pool ball tracking.
Most of cases, your method works well.
Sometimes, but, after two balls collide, the IDs are  exchanged.

![image](https://github.com/tryolabs/norfair/assets/135291100/cb47ab7e-10b8-4ddd-b268-46b90c116415)
![image](https://github.com/tryolabs/norfair/assets/135291100/1cc75995-19a5-4f6d-9692-cd5d805c86f3)
![image](https://github.com/tryolabs/norfair/assets/135291100/afb47574-8e48-4bc5-b598-e7a4e488f9e5)
![image](https://github.com/tryolabs/norfair/assets/135291100/e10e3474-83d7-4b41-8c96-f6a6a2a2f22c)

**Comment by aguscas** [comment]
Hello! This is a common problem in tracking, but I have a few suggestions you can try. The main reason this problem happens is because we estimate the velocities of the objects when tracking them, to predict where they should be in the following frame. The sudden change in velocity during these collisions makes the tracking more challenging.

These are a couple things you might want to try:

1. **Don't use the Kaman filter**: When initializing your tracker, you can set the argument `filter_factory` to `NoFilterFactory()` to not use the velocity. 
```python
from norfair.filter import NoFilterFactory

 tracker = Tracker(
    distance_function=distance_function, 
    detection_threshold=DETECTION_THRESHOLD,  
    distance_threshold=DISTANCE_THRESHOLD,
    hit_counter_max=HIT_COUNTER_MAX,
    initialization_delay=INITIALIZATION_DELAY,
    pointwise_hit_counter_max=POINTWISE_HIT_COUNTER_MAX,
    filter_factory=NoFilterFactory(),    # with this line, you wouldn't be using the velocity
)
```
Remember that doing this might cause other problems when the balls are moving normally, since we just use the last position of the ball (and not its velocity) to estimate where it should be on the following frame. You might want to change your `distance_function` and `DISTANCE_THRESHOLD` for that. I don't know which distance function you are currently using, but for example, I would suggest using the euclidean distance between the centers of the balls instead of IoU, since the boxes of the predicted position and the detected position might not even overlap if a ball moves to fast. If you need any help with this, please let me know.


2. You can try using a **distance function that considers an embedding** of how the object looks like. So in a similar fashion (but not exactly the same) to what it was done in the[ Reid demo](https://github.com/tryolabs/norfair/tree/master/demos/reid), you can compare how similar two objects look instead of just their position on the screen. 

So, for example, for each detection you could crop the bounding box and compute its [color histogram](https://www.geeksforgeeks.org/python-opencv-cv2-calchist-method/) with opencv and normalize it. 
```python
def get_hist(image):
    hist = cv2.calcHist(
        [cv2.cvtColor(image, cv2.COLOR_BGR2Lab)],
        [0, 1],
        None,
        [128, 128],
        [0, 256, 0, 256],
    )
    return cv2.normalize(hist, hist).flatten()
```

and put that in the `embedding` attribute of the detection

```python
for detection in detections:
  cut = get_cutout(detection.points, frame) # get_cutout would be a function that crops the bounding box
  if cut.shape[0] > 0 and cut.shape[1] > 0:
      detection.embedding = get_hist(cut)
  else:
      detection.embedding = None
```

Then you can define a custom `distance_function` that compares the embedding of the detection in the current frame, with the embedding of the `last_detection` associated to a tracked object. You can even combine both the spatial distance and the similarity of the embeddings, either by simply multiplying or adding both values, or for example doing conditionals (if the objects look similar, then you compute the spatial distance and use that as the final distance, otherwise if the objets don't look then set that distance to a really large number so that they won't ever match). The following is an example just using the embedding distance:

```python
def your_custom_distance_function(detection: "Detection", tracked_object: "TrackedObject") -> float:
  for past_detection in reversed(tracked_object.past_detections):
      if past_detection.embedding is not None:
          last_embedding = past_detection.embedding
          break
  else:
      return 1

  # this should give a distance between 0 and 1, since we are just doing 1-correlation 
  distance = 1 - cv2.compareHist(
      last_embedding, detection.embedding, cv2.HISTCMP_CORREL
  ) 
  return distance
```

Please note that I haven't tested any code snippet here, so there might be some problems to address when trying to run them. If you need any further assistance, please let me know.

**Comment by rose-jinyang** [comment]
Thanks for your full guidance.

**Comment by armine105** [comment]
Hi @aguscas
Thank you very much
I fixed the above issue by revising the mean_eculidean function as the following.

![image](https://github.com/tryolabs/norfair/assets/135291100/6bd3ca21-d75c-4e2f-9a7f-2224d7643aaf)

Most of cases, it works well.
That is, when associating a detection and an already existing tracked object, I used the tracked object's last detection position rather than the estimated next position.
